tiledb/sm/metadata/metadata.cc

```diff
@@ -210,8 +210,15 @@ void Metadata::store(
   resources.stats().add_counter(
       "write_meta_size", size_computation_serializer.size());
 
-  // Create a metadata file name
-  GenericTileIO::store_data(resources, get_uri(uri), tile, encryption_key);
+  if (uri.is_file()) {
+    // For LocalFS use a .tmp extension until write is flushed.
+    URI metadata_uri(get_uri(uri).to_string() + ".tmp");
+    GenericTileIO::store_data(resources, metadata_uri, tile, encryption_key);
+    resources.vfs().move_file(metadata_uri, get_uri(uri));
+  } else {
+    // Create a metadata file name
+    GenericTileIO::store_data(resources, get_uri(uri), tile, encryption_key);
+  }
```

Comment:
Not only metadata is susceptible to these race conditions; I can think of group members, array schema evolution, and array creation. It's OK if we want to prioritize metadata due to the increased impact.

I would have preferred to move this temporary file logic to the VFS, which would have required a pretty significant refactoring, but doing it this way is not very disruptive and OK for now.